### ISSUE TITLE
runtime: new goroutines can spend excessive time in morestack

### ISSUE THREAD

**Issue by petermattis** [golang/go#18138]
### What version of Go are you using (`go version`)?

`go version devel +41908a5 Thu Dec 1 02:54:21 2016 +0000 darwin/amd64` a.k.a `go1.8beta1`

### What operating system and processor architecture are you using (`go env`)?

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/pmattis/Development/go"
GORACE=""
GOROOT="/Users/pmattis/Development/go-1.8"
GOTOOLDIR="/Users/pmattis/Development/go-1.8/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/go-build385423377=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
```

### What did you do?

A recent [change](https://github.com/cockroachdb/cockroach/pull/11196) to [github.com/cockroachdb/cockroach](http://github.com/cockroachdb/cockroach) replaced a synchronous call with one wrapped in a goroutine. This small change resulted in a significant slowdown in some benchmarks. The slowdown was traced to additional time being spent in `runtime.morestack`. The problematic goroutines are all hitting a single gRPC entrypoint [`Server.Batch`](https://github.com/cockroachdb/cockroach/blob/master/pkg/server/node.go#L845) and the code paths that fan out from this entrypoint tend to use an excessive amount of stack due to an over reliance on passing and returning by value instead of using pointers. Typical calls use 16-32 KB of stack.

The expensive part of `runtime.morestack` is the adjustment of existing pointers on the stack. And due to the incremental nature of the stack growth, I can see the stack growing in 4 steps from 2 KB to 32 KB. So we experimented with a hack to pre-grow the stack. Voila, the performance penalty of the change disappeared:

```
name               old time/op  new time/op  delta
KVInsert1_SQL-8     339µs ± 2%   312µs ± 1%   -7.89%  (p=0.000 n=10+10)
KVInsert10_SQL-8    485µs ± 2%   471µs ± 1%   -2.81%  (p=0.000 n=10+10)
KVInsert100_SQL-8  1.36ms ± 0%  1.35ms ± 0%   -0.95%  (p=0.000 n=10+10)
KVUpdate1_SQL-8     535µs ± 1%   487µs ± 1%   -9.02%   (p=0.000 n=10+9)
KVUpdate10_SQL-8    777µs ± 1%   730µs ± 1%   -6.03%   (p=0.000 n=10+9)
KVUpdate100_SQL-8  2.69ms ± 1%  2.66ms ± 1%   -1.16%  (p=0.000 n=10+10)
KVDelete1_SQL-8     479µs ± 1%   429µs ± 2%  -10.43%   (p=0.000 n=9+10)
KVDelete10_SQL-8    676µs ± 1%   637µs ± 1%   -5.80%    (p=0.000 n=9+9)
KVDelete100_SQL-8  2.23ms ± 5%  2.18ms ± 4%     ~     (p=0.105 n=10+10)
KVScan1_SQL-8       216µs ± 5%   179µs ± 1%  -17.12%  (p=0.000 n=10+10)
KVScan10_SQL-8      233µs ± 1%   201µs ± 1%  -13.76%  (p=0.000 n=10+10)
KVScan100_SQL-8     463µs ± 1%   437µs ± 0%   -5.64%   (p=0.000 n=10+8)
```

`old` are benchmarks gathered using go1.8beta1 and `new` are on go1.8beta1 with the hack to pre-grow the stack. The hack is a call at the beginning of `server.Batch` to a `growStack` method:

```
var growStackGlobal = false

//go:noinline
func growStack() {
	// Goroutine stacks currently start at 2 KB in size. The code paths through
	// the storage package often need a stack that is 32 KB in size. The stack
	// growth is mildly expensive making it useful to trick the runtime into
	// growing the stack early. Since goroutine stacks grow in multiples of 2 and
	// start at 2 KB in size, by placing a 16 KB object on the stack early in the
	// lifetime of a goroutine we force the runtime to use a 32 KB stack for the
	// goroutine.
	var buf [16 << 10] /* 16 KB */ byte
	if growStackGlobal {
		// Make sure the compiler doesn't optimize away buf.
		for i := range buf {
			buf[i] = byte(i)
		}
	}
}
```

The question here is whether this is copacetic and also to alert the runtime folks that there is a performance opportunity here. Note that the `growStackGlobal` is not currently necessary, but I wanted to future proof against the compiler deciding that `buf` is not necessary. 

Longer term, the stack usage under `server.Batch` should be reduced on our side. I'm guessing that we could get the stack usage down to 8-16 KB without too many contortions. But even with such reductions, being able to pre-grow the stack for a goroutine looks beneficial.

**Comment by aclements** [comment]
We've seen this a few times now. I'm not sure what the right answer is. My best thought so far is that the runtime could keep track of when particular go statements always lead to stack growth right away (for some value of "right away" and "always") and learn to start goroutines from that site with a larger stack. Of course, it would be hard to make this behavior predictable, but perhaps it would still be less surprising than the current behavior. If the runtime did learn to start a goroutine with a larger stack, it would still need a signal to learn if the stack should get smaller again, but we could do that efficiently by allocating the larger stack but setting the stack bounds to something smaller. Then the runtime could still observe whether or not the stack needs to grow, but the actual growth would be basically free until it reached the size of the allocation.

@randall77, thoughts, ideas?

/cc @RLH

**Comment by mrjrieke** [comment]
I like @petermattis idea of being able to hint stack size on a per goroutine basis, although this implies developers have the knowhow to identify and provide size estimates accurately.  Could this be done with a compiler directive?

**Comment by bradfitz** [comment]
We don't want compiler directives in code. We have some used by the runtime out of necessity, but they're gross. Go prefers simplicity over tons of knobs.

**Comment by petermattis** [comment]
Yes, please just make my code magically faster as you've been doing for the last several Go releases.

**Comment by mrjrieke** [comment]
I generally agree with not having compiler directives ... magic is nice, although they (compiler directives) do exist even in go.  It's an interesting opportunity either way you decide.

**Comment by mrjrieke** [comment]
@bradfitz, your comment prompted me to look for the go guiding principles ( https://golang.org/doc/faq#principles).  Thanks @adg as well for nicely worded principles.

**Comment by aclements** [comment]
@petermattis (or anyone who has a good reproducer for this), would you be able to try https://go-review.googlesource.com/45142? It's a trivial hack, but it might actually do the trick. I haven't benchmarked it on anything, so it may also slow things down.

**Comment by petermattis** [comment]
@aclements I'll try and test either tomorrow or next week.

**Comment by petermattis** [comment]
@aclements Applying that patch to go1.8.3 resulted in no benefit (this is with the `growStack` hack disabled):

```
~/Development/go/src/github.com/cockroachdb/cockroach/pkg/sql master benchstat out.old out.new
name                old time/op  new time/op  delta
KV/Insert1_SQL-8     363µs ± 3%   369µs ± 2%  +1.43%  (p=0.043 n=10+10)
KV/Insert10_SQL-8    583µs ± 0%   581µs ± 1%    ~     (p=0.113 n=10+9)
KV/Insert100_SQL-8  2.05ms ± 0%  2.05ms ± 1%    ~     (p=0.912 n=10+10)
KV/Update1_SQL-8     578µs ± 1%   577µs ± 1%    ~     (p=0.968 n=9+10)
KV/Update10_SQL-8    913µs ± 1%   914µs ± 1%    ~     (p=0.931 n=9+9)
KV/Update100_SQL-8  3.80ms ± 1%  3.87ms ± 5%  +1.90%  (p=0.019 n=10+10)
KV/Delete1_SQL-8     517µs ± 2%   518µs ± 2%    ~     (p=0.912 n=10+10)
KV/Delete10_SQL-8    813µs ± 2%   809µs ± 1%    ~     (p=0.280 n=10+10)
KV/Delete100_SQL-8  3.22ms ± 2%  3.26ms ± 3%    ~     (p=0.052 n=10+10)
KV/Scan1_SQL-8       217µs ± 1%   216µs ± 0%    ~     (p=0.090 n=9+10)
KV/Scan10_SQL-8      238µs ± 0%   238µs ± 1%    ~     (p=0.122 n=10+8)
KV/Scan100_SQL-8     454µs ± 0%   455µs ± 1%    ~     (p=0.089 n=10+10)
```

Surprising to me this didn't have any effect. Compare this to the `growStack` hack mentioned earlier:

```
~/Development/go/src/github.com/cockroachdb/cockroach/pkg/sql master benchstat out.old out.grow-stack
name                old time/op  new time/op  delta
KV/Insert1_SQL-8     363µs ± 3%   331µs ± 2%   -8.82%  (p=0.000 n=10+10)
KV/Insert10_SQL-8    583µs ± 0%   561µs ± 1%   -3.80%  (p=0.000 n=10+10)
KV/Insert100_SQL-8  2.05ms ± 0%  2.03ms ± 0%   -0.88%  (p=0.000 n=10+8)
KV/Update1_SQL-8     578µs ± 1%   532µs ± 1%   -7.94%  (p=0.000 n=9+10)
KV/Update10_SQL-8    913µs ± 1%   872µs ± 1%   -4.47%  (p=0.000 n=9+9)
KV/Update100_SQL-8  3.80ms ± 1%  3.75ms ± 1%   -1.36%  (p=0.000 n=10+10)
KV/Delete1_SQL-8     517µs ± 2%   458µs ± 2%  -11.54%  (p=0.000 n=10+10)
KV/Delete10_SQL-8    813µs ± 2%   765µs ± 1%   -5.91%  (p=0.000 n=10+10)
KV/Delete100_SQL-8  3.22ms ± 2%  3.16ms ± 1%   -2.01%  (p=0.000 n=10+10)
KV/Scan1_SQL-8       217µs ± 1%   194µs ± 1%  -10.44%  (p=0.000 n=9+10)
KV/Scan10_SQL-8      238µs ± 0%   216µs ± 1%   -9.36%  (p=0.000 n=10+10)
KV/Scan100_SQL-8     454µs ± 0%   431µs ± 1%   -4.92%  (p=0.000 n=10+9)
```

**Comment by josharian** [comment]
CL 43150 might help a little here.

**Comment by aclements** [comment]
Sorry, I made a silly mistake in CL 45142. Would you mind trying the new version of that CL?

**Comment by petermattis** [comment]
With your updated patch against go-tip (f363817f143a7cee0ccdb7833a99ac1e8f210197) there is an improvement:

```
~/Development/go/src/github.com/cockroachdb/cockroach/pkg/sql master benchstat out.old out.new
name              old time/op  new time/op  delta
KV/Scan1_SQL-8     243µs ± 1%   224µs ± 0%  -7.57%  (p=0.000 n=9+9)
KV/Scan10_SQL-8    263µs ± 0%   247µs ± 0%  -6.20%  (p=0.000 n=9+10)
KV/Scan100_SQL-8   463µs ± 0%   444µs ± 0%  -4.05%  (p=0.000 n=10+10)
```

But the improvement is still not as good as the `growStack` hack:

```
~/Development/go/src/github.com/cockroachdb/cockroach/pkg/sql master benchstat out.new out.grow-stack
name              old time/op  new time/op  delta
KV/Scan1_SQL-8     224µs ± 0%   219µs ± 0%  -2.24%  (p=0.000 n=9+9)
KV/Scan10_SQL-8    247µs ± 0%   240µs ± 1%  -2.59%  (p=0.000 n=10+10)
KV/Scan100_SQL-8   444µs ± 0%   439µs ± 0%  -1.06%  (p=0.000 n=10+9)
```

There is a little more performance if we increase the initial stack size to 32 KB:

```
~/Development/go/src/github.com/cockroachdb/cockroach/pkg/sql master benchstat out.old out.new2
name              old time/op  new time/op  delta
KV/Scan1_SQL-8     243µs ± 1%   209µs ± 1%  -13.76%  (p=0.000 n=9+9)
KV/Scan10_SQL-8    263µs ± 0%   232µs ± 2%  -11.61%  (p=0.000 n=9+10)
KV/Scan100_SQL-8   463µs ± 0%   445µs ± 4%   -3.86%  (p=0.000 n=10+9)
```

Interestingly, all of these timings are lower than with go1.8.3.

**Comment by petermattis** [comment]
> Interestingly, all of these timings are lower than with go1.8.3.

Nothing to see here. This appears to be due to a change on our code between what I tested earlier today and now.

**Comment by petermattis** [comment]
I did some more testing of this patch and the performance improvements carry over to production settings. `morestack` disappears from profiles. Note this is using a version of the patch which uses a 32KB initial stack size.

**Comment by petermattis** [comment]
It is early in the 1.10 cycle and wanted to bring this issue forward again. See https://github.com/cockroachdb/cockroach/issues/17242 for a graph showing the benefit of a larger initial stack size.

**Comment by petermattis** [comment]
Is there any update on this issue? A larger initial goroutine stack size provides a nice performance boost for our system.

**Comment by richardartoul** [comment]
@petermattis We ran into a similar issue in production where adding a single extra function call (few lines of code, nothing heavy) increased our end-to-end latency by a few seconds. We narrowed down the issue to an increase in runtime.morestack. We ended up resolving the issue with this worker pool: https://github.com/m3db/m3x/blob/master/sync/pooled_worker_pool.go

I noticed that you said:

> I can see the stack growing in 4 steps from 2 KB to 32 KB.

Did you have any way of programatically measuring when the stack grows? I would love to be able to prove in production that the additional function we added was enough to push the stack over the edge into one more copy, but all I could do was make inferences based on the CPU profiles.

**Comment by josharian** [comment]
If you can reproduce in a non-production environment, you might want to try tip of Go 1.11. There were a few improvements made this cycle that could help a bit.

Again, in a non-production environment, you could simply add a println(“grow”) to the implementation of morestack in the runtime. In a production environment, you might want to atomically increment a package variable instead and read it out atomically in another goroutine.

You might also want to file an issue requesting a built-in pprof profile for stack size changes, although even if approved that couldn’t land until 1.12.

**Comment by richardartoul** [comment]
@josharian I'm happy to request a built-in pprof profile for stack size changes if you think that's possible. The main issue with the current CPU profiles is that you can see its happening, but can't trace what is triggering it

**Comment by josharian** [comment]
Please file an issue, and we can discuss feasibility there. If you cc me I’ll pull in other relevant folks. Thanks!

**Comment by richardartoul** [comment]
@petermattis Awesome, thanks! Will report back.

@josharian I will file an issue soon, just want to gather all my evidence as its kind of scattered around right now.

**Comment by josharian** [comment]
Don’t worry about evidence or a big write up. For this, a few sentences and a reference back to this issue suffices.

**Comment by CAFxX** [comment]
Just a note: I have a branch where I started implementing PC-based stack estimation. Basically, when a G is created it checks if we have an estimate for the most common stack size used by Gs started at the same PC. 
It shows potential but it's pretty crude because it just measures stack *size* instead of actual stack *utilization* (estimates are reset during GC to prevent infinite growth). Measuring (or at least *estimating*) actual stack utilization would likely require modifying the `morestack` machinery (but that seems to involve modifying a lot of assembler code of platforms I'm not familiar with). It would also benefit from compiler support, but that's less important.
If anybody wants to give it a spin, it's here: https://github.com/golang/go/compare/master...CAFxX:dynstack-local
By default stack estimation is enabled. Running the built programs using `GOSTACKESTIM=0` disables stack estimation (i.e. the normal behavior). `GOSTACKESTIMDEBUG=1` will print a line every time a goroutine is started with a non-standard stack size as a result of stack estimation.

**Comment by aclements** [comment]
For reference, this issue was just discussed at length in an [Uber Engineering blog post](https://eng.uber.com/optimizing-m3/) (issue #31558 pointed that out).

**Comment by CAFxX** [comment]
@aclements regarding this issue, and specifically my experiment above, I ran into issue while implementing stack estimation as the code currently implicitly assumes that the stack guard size is constant in many different points throughout the runtime.

This is more of a wish than anything since you are reworking exactly that (https://go-review.googlesource.com/c/go/+/172284/) but... if you could "centralize" the assumption so that it's not spread around the runtime, I would be in a much better position to wrap up my attempt at fixing this.

The reason I need this is that to estimate stack usage when using largen-than-normal stacks I allocate a stack of size N and set the stack guard so that morestack is called when more than N/2 is in use, at which point the stack guard reverts to the default. When the G dies I check if the stack guard is still at N/2 or if it's default; in the former case I know the G never used more than half the estimated space, and I can adjust the estimation accordingly.

**Comment by lrita** [comment]
In my service, we using the classical server mode, which spawn a goroutine by a new connection. There are many different clients, someone use long term connection, and the others use a short term connection, and we cannot optimize those short term clients. 

We use snappy compressing for network communicating. The [snappy.encodeBlock](https://github.com/golang/snappy/blob/master/encode_amd64.s#L265) need a 32k stack, that spend excessive time in morestack for short term connection clients.

In practice, we does not want wait a long long time for a good, final solution. I think compiler directives is a simply and fast way to resolve these relevant performance issue, although it is gross. We can benefit from it soon.

**Comment by valyala** [comment]
I believe high stack usage in `snappy.encodeBlock` must be fixed, since it results in high memory usage in apps serving high number of concurrent connections. Stack hog functions could be detected with the [proposed stack profiler](https://github.com/golang/go/issues/20010).

**Comment by marcogrecopriolo** [comment]
Just a heads up that this is an issue for Couchbase's N1QL as well.
Personally, a directive would work well for N1QL (we only have one entry point where we need to run with a different stack size), but I understand the reluctance to use explicit controls.
OTOH - having to build a whole infrastructure of goroutine workers to seems a high price to pay for one very small ideological sin (one directive).
Maybe we could have a little map of frequently used entry points holding each point's average stack size on exit?

**Comment by uluyol** [comment]
Not sure if this issue was noted anywhere. Here is a sample program that has enough static information to avoid any calls to morestack but in fact observes multiple stack growths:

```
package main

var shouldSet = false
var c = make(chan bool)
var x [16384]byte

func main() {
	go f32()
	<-c
	println("exit")
}

//go:noinline
func f32() {
	var buf [32]byte
	if shouldSet {
		// Make sure the compiler doesn't optimize away buf.
		for i := range buf {
			buf[i] = byte(i)
		}
		copy(x[:], buf[:])
	}
	f64()
}

[...]


//go:noinline
func f16384() {
	var buf [16384]byte
	if shouldSet {
		// Make sure the compiler doesn't optimize away buf.
		for i := range buf {
			buf[i] = byte(i)
		}
		copy(x[:], buf[:])
	}
	println("done")
	c <- true
}
```

Each function fX allocates a variable of X bytes on the stack and unconditionally calls another function f(2*X) up to 16KB. When `go f32()` is called, the compiler should have enough information to allocate a large-enough stack frame up front (it knows that f32 needs to allocate stack space for f64, f128, and so on). What we see instead are multiple calls to newstack:

<details>

```
runtime: newstack: 2048 -> 4096
runtime.(*mcentral).grow(0x10f6b98, 0x0)
	/Users/m/dev/go/src/runtime/mcentral.go:264 +0x13d
runtime.(*mcentral).cacheSpan(0x10f6b98, 0x1402200)
	/Users/m/dev/go/src/runtime/mcentral.go:106 +0x2bc
runtime.(*mcache).refill(0x1121108, 0xe)
	/Users/m/dev/go/src/runtime/mcache.go:138 +0x84
runtime.(*mcache).nextFree(0x1121108, 0xe, 0x1402300, 0x15fffff, 0xc00002e5c0)
	/Users/m/dev/go/src/runtime/malloc.go:867 +0x87
runtime.mallocgc(0x60, 0x10700e0, 0xc00002e601, 0x49)
	/Users/m/dev/go/src/runtime/malloc.go:1047 +0x792
runtime.newobject(0x10700e0, 0x100a024)
	/Users/m/dev/go/src/runtime/malloc.go:1176 +0x38
runtime.acquireSudog(0xc000044048)
	/Users/m/dev/go/src/runtime/proc.go:344 +0x281
runtime.chanrecv(0xc000044000, 0x0, 0xc000000101, 0x101204e)
	/Users/m/dev/go/src/runtime/chan.go:551 +0x223
runtime.chanrecv1(0xc000044000, 0x0)
	/Users/m/dev/go/src/runtime/chan.go:433 +0x2b
runtime.gcenable()
	/Users/m/dev/go/src/runtime/mgc.go:216 +0x95
runtime.main()
	/Users/m/dev/go/src/runtime/proc.go:167 +0x115
runtime.goexit()
	/Users/m/dev/go/src/runtime/asm_amd64.s:1374 +0x1
runtime: newstack: 2048 -> 4096
main.f512()
	/Users/m/dev/go/nomorestack/main.go:75 +0x86
main.f256()
	/Users/m/dev/go/nomorestack/main.go:62 +0x86
main.f128()
	/Users/m/dev/go/nomorestack/main.go:49 +0x83
main.f64()
	/Users/m/dev/go/nomorestack/main.go:36 +0x80
main.f32()
	/Users/m/dev/go/nomorestack/main.go:23 +0x5e
created by main.main
	/Users/m/dev/go/nomorestack/main.go:8 +0x35
runtime: newstack: 4096 -> 8192
main.f1024()
	/Users/m/dev/go/nomorestack/main.go:88 +0x91
main.f512()
	/Users/m/dev/go/nomorestack/main.go:75 +0x86
main.f256()
	/Users/m/dev/go/nomorestack/main.go:62 +0x86
main.f128()
	/Users/m/dev/go/nomorestack/main.go:49 +0x83
main.f64()
	/Users/m/dev/go/nomorestack/main.go:36 +0x80
main.f32()
	/Users/m/dev/go/nomorestack/main.go:23 +0x5e
created by main.main
	/Users/m/dev/go/nomorestack/main.go:8 +0x35
runtime: newstack: 8192 -> 16384
main.f2048()
	/Users/m/dev/go/nomorestack/main.go:101 +0x81
main.f1024()
	/Users/m/dev/go/nomorestack/main.go:88 +0x91
main.f512()
	/Users/m/dev/go/nomorestack/main.go:75 +0x86
main.f256()
	/Users/m/dev/go/nomorestack/main.go:62 +0x86
main.f128()
	/Users/m/dev/go/nomorestack/main.go:49 +0x83
main.f64()
	/Users/m/dev/go/nomorestack/main.go:36 +0x80
main.f32()
	/Users/m/dev/go/nomorestack/main.go:23 +0x5e
created by main.main
	/Users/m/dev/go/nomorestack/main.go:8 +0x35
runtime: newstack: 16384 -> 32768
main.f4096()
	/Users/m/dev/go/nomorestack/main.go:114 +0x97
main.f2048()
	/Users/m/dev/go/nomorestack/main.go:101 +0x81
main.f1024()
	/Users/m/dev/go/nomorestack/main.go:88 +0x91
main.f512()
	/Users/m/dev/go/nomorestack/main.go:75 +0x86
main.f256()
	/Users/m/dev/go/nomorestack/main.go:62 +0x86
main.f128()
	/Users/m/dev/go/nomorestack/main.go:49 +0x83
main.f64()
	/Users/m/dev/go/nomorestack/main.go:36 +0x80
main.f32()
	/Users/m/dev/go/nomorestack/main.go:23 +0x5e
created by main.main
	/Users/m/dev/go/nomorestack/main.go:8 +0x35
runtime: newstack: 32768 -> 65536
main.f8192()
	/Users/m/dev/go/nomorestack/main.go:127 +0x97
main.f4096()
	/Users/m/dev/go/nomorestack/main.go:114 +0x97
main.f2048()
	/Users/m/dev/go/nomorestack/main.go:101 +0x81
main.f1024()
	/Users/m/dev/go/nomorestack/main.go:88 +0x91
main.f512()
	/Users/m/dev/go/nomorestack/main.go:75 +0x86
main.f256()
	/Users/m/dev/go/nomorestack/main.go:62 +0x86
main.f128()
	/Users/m/dev/go/nomorestack/main.go:49 +0x83
main.f64()
	/Users/m/dev/go/nomorestack/main.go:36 +0x80
main.f32()
	/Users/m/dev/go/nomorestack/main.go:23 +0x5e
created by main.main
	/Users/m/dev/go/nomorestack/main.go:8 +0x35
done
exit
```

</details>

Unfortunately, I don't have any idea of how much fixing this would help in practice.

**Comment by marcogrecopriolo** [comment]
Is there any way https://golang.org/cl/225800 could be merged in 1.14.x?

In between pooling goroutines and growing the stack with a dummy function call, the second is the better option (too much contention operating the goroutine queues), however guessing the stack is a bit of a dark art: too small and you achieve nothing, too big and you get a huge bottleneck in stackcacherefill().

If we could allocate a nice round number in one go without having to worry about if it is slightly too big, it would make our life easier: at least we don't have to experiment to find the right stack size.

Also - if you want numbers from a real life high throughput service: untreated, this is issue is eating up 12% of our overall CPU time.
Hacking the stack grows our throughput by 6% and still needlessly uses 6% CPU time.
Pooling goroutines on a partitioned queue (as many fragments as cores) removes the newstack() CPU load, but the contention in managing the queues is such that we improve throughput by 1% or thereabout.

Starting the goroutine with a larger stack size would be best.

**Comment by aclements** [comment]
@marcogrecopriolo, I'm not sure I quite understand:

> Hacking the stack grows our throughput by 6% and still needlessly uses 6% CPU time.

Do you mean that CL 225800 increases your throughput by 6%, or something else?

> Also - if you want numbers from a real life high throughput service: untreated, this is issue is eating up 12% of our overall CPU time.

That is interesting. Can you tell from profile stacks that this is definitely happening right after new goroutines start?

> Starting the goroutine with a larger stack size would be best.

This is a very tricky trade-off, since many systems that create large numbers of goroutines benefit from the small starting stacks. There's no simple answer here, I'm afraid. :( CL 225800 has basically no downside, so I would love to know if it actually does help.

**Comment by marcogrecopriolo** [comment]
What I mean is that if I use code like this to fire my goroutine:

`// MB-38469 / go issue 18138 initial goroutine stack too small
//go:noinline
func primeStack() {
	const _STACK_BUF_SIZE = 512 // 128 multiples
	var buf [_STACK_BUF_SIZE]int64

	// force the compiler to allocate buf
	for i := 127; i < _STACK_BUF_SIZE; i += 128 {
		buf[i] = int64(i)
	}

	_ = stackTop(buf[_STACK_BUF_SIZE-1])
}

//go:noinline
func stackTop(v int64) int64 {
	return v
}

func execOp(op Operator, context *Context, parent value.Value) {
	primeStack()
	op.RunOnce(context, parent)
}

// fork operator
func (this *base) fork(op Operator, context *Context, parent value.Value) {
	if op.getBase().inline {
		this.switchPhase(_NOTIME)
		op.RunOnce(context, parent)
		this.switchPhase(_EXECTIME)
	} else {
		go execOp(op, context, parent)
		// go op.RunOnce(context, parent)
	}
}`

(go op.RunOnce(context, parent) is how I would normally call it), and size STACK_BUF_SIZE roughly speaking right, I get better throughput, and lower the cost of newstack().
In my case 512 is giving me +6% throughput and half the cost in newstack().

If I get STACK_BUF_SIZE wrong, I either get a huge bottleneck in stackcacherefill() and a drop in throughput. In my case STACK_BUF_SIZE set to 1024 gets me a 10% throughput drop and newstack() taking twice as much CPU time, mostly spent in stackcacherefill() piling up in some lock.

I would need to spend more time finding the optimal value for STACK_BUF_SIZE (I know it should be lower than 512), but it takes me more than an hour to run my performance test rig, and even then, somebody might actually increase their stack need down the line and render my testing useless.

If I could have 225800 merged, then I wouldn't need to worry about contention in stackcacherefill(), I'd just size my stack to a reasonable amount and never worry about it ever again.

As a side note, I know we want to avoid compilers directives, but all of the above is just a poor man's way to say

`//go:stack=4196
go op.RunOnce(context, parent)`

Which would be cleaner.
That would give the option of creating goroutines with larger stacks, where specifically required and when growing stacks is known to be a cost, and systems that create large numbers of goroutines would still benefit from small stacks as usual.

**Comment by marcogrecopriolo** [comment]
Sorry - I should clarify, from extensive profiling I know that newstack() is only called in primestack(), so yes it only happens at goroutine starts.

**Comment by roudkerk** [comment]
Rather than a compiler directive, would it be possible to add functions to the `runtime` package

```
GoWithStackHint(f func(), stackSize int)
StackSize() int
```

The first could be used to start goroutines with statically determined stack size, for example

```
runtime.GoWithStackHint(func() { op.RunOnce(context, parent) }, 4096)
```

But it would also be possible to write a helper that dynamically chooses stack size based on the history of final stack sizes for goroutines created with that helper instance.  For example

```
var runner = smartstacksize.New()

...

  runner.Go(func() { op.RunOnce(context, parent) })
```

**Comment by CAFxX** [comment]
Forgot to mention it back here, but I have a [CL](https://golang.org/cl/341990) up for early review that basically implements something [similar to what @aclements suggested](https://github.com/golang/go/issues/18138#issuecomment-264228318). It uses a fairly conservative approach, so in this early incarnation it may still leave a bit of performance on the table, but in the tests so far it seems to work well enough in practice to be useful. If anyone wants to run their own benchmarks and report back it would be great (you need to build from that CL and then set `GOEXPERIMENT=predictstacksize`). The upside is that it requires no knobs, annotations, or code changes.

**Comment by randall77** [comment]
I wrote up an idea I had about starting goroutines with a larger than minimum stack size, for this issue. The doc is [here](https://docs.google.com/document/d/1YDlGIdVTPnmUiTAavlZxBI1d9pwGQgZT7IKFKlIXohQ/edit?usp=sharing).
The immediate impetus for this doc was another attempt to fix that issue in [CL 341990](https://go-review.googlesource.com/c/go/+/341990/), but generally these ideas have been sloshing around my head for a while.
Comments welcome. I have a first stab at an implementation in [CL 345889](https://go-review.googlesource.com/c/go/+/345889).

**Comment by go101** [comment]
Improve @uluyol's [idea](https://github.com/golang/go/issues/18138#issuecomment-605132847) by setting the initial stack size of a new goroutine to any 2<sup>n</sup> sizes:

```Go
func startRoutine() {
	// Use a dummy anonymous function to enlarge stack.
	func(x *interface{}) {
		type _ int // avoid being inlined
		if x != nil {
			*x = [128 << 20]byte{} // initial 256M stack
		}
	}(nil)
	
	// ... do work load
}
```

**Comment by go101** [comment]
> Rather than a compiler directive, would it be possible to add functions to the runtime package

It would be great to add a `runtime/debug/SetCurrentGoroutineOption` function:
```Go
type GoroutineOption int
func SetCurrentGoroutineOption(key GoroutineOption, value int) {...}

const (
	StackSizeOfTheNextSpawndGoroutine GoroutineOption = iota
	PriorityCasesInTheNextSelectBlock
	...
)
```

**Comment by pcostanza** [comment]
It would be good if Go would support tail call elimination. Then at least in some cases, the stack wouldn't grow as much as it does now.

**Comment by ianlancetaylor** [comment]
A straightforward implementation of tail call elimination would lead to difficulties with `runtime.Caller` and `runtime.Callers`.

See also #22624.

**Comment by pcostanza** [comment]
"Proper" tail recursion requires the full removal of the caller's stack frame when a function is call in tail position. If that is too extreme, maybe it's possible to only remove the part of the stack frame that is not needed for `runtime.Caller` and `runtime.Callers`, in the hope to reduce time spent in morestack?

**Comment by go101** [comment]
It looks a new implementation will be adopted in Go 1.19.
After reading the design doc of the change, it looks now a global average stack size is used as the initial stack size for new spawn goroutines. Will this be efficient if the variance of goroutine stack sizes is large?

**Comment by randall77** [comment]
Yes. Of course, it matters what exactly you mean by "efficient". The new behavior uses a bit more space for a bit less CPU.

Starting new goroutines with the average stack size will waste at most a factor of 2 in space (in addition to the at-most factor of 2 we already waste by growing stacks using power-of-two sizes). In exchange, we get less work on goroutine startup.

It is conceivable that the tradeoff is not advantageous for some programs. I suspect it will be a net win for most. I'm curious if anyone has an example where it is not the right thing to do.

**Comment by go101** [comment]
I'm a little worrying about this change will bring more unpredictable/random factors for Go program performances.

**Comment by SaveTheRbtz** [comment]
Is this still a problem after [go 1.19](https://tip.golang.org/doc/go1.19)?

> The runtime will now allocate initial goroutine stacks based on the historic average stack usage of goroutines. This avoids some of the early stack growth and copying needed in the average case in exchange for at most 2x wasted space on below-average goroutines.

**Comment by randall77** [comment]
I have not heard any complaints.

**Comment by marcogrecopriolo** [comment]
FWIW [historic average stack usage](https://go-review.googlesource.com/c/go/+/345889) makes absolutely no difference to   the Couchbase's N1QL engine.
We still have one class of goroutines (the execution layer operators) that would benefit from having an initial stack size which is much larger than the historic stack size, and I still have to hack an initial stack allocation.
So I am basically fudging the code adding extra cost just to have a lower cost for initial stack allocation.

For us, if a compiler directive is not acceptable, at least having some compiler analysis that determines the initial stack size for a goroutine (you have sample code that you can use for testing from myself and other contributors) would work much better than the status quo.

**Comment by marcogrecopriolo** [comment]
FWIW I believe [average stack sizes](https://go-review.googlesource.com/c/go/+/345889) not to be a great approach, and [stack sizes based on initial PC](https://golang.org/cl/341990) would have been much better.

In a classical server example, where each request is serviced by one or more goroutines, that stack size is not a random occurrence, but is a function of the work done by the individual goroutine.

In our case, all execution operators use a lot of stack, because they do the bulk of the work, everything else does not.
What the chosen approach achieves is to give consistently more stack to those goroutines that don't needed and less stack to those that do, ie it fixes nothing, while at the same time wasting stack.

To put this in context, on my performance testing cluster, we serve about 200K requests a second, and spawn 5 execution operators, which end up having to grow the stack about 1M times a second.
Requests take about 0.7ms from admission to completion.

Also, to put this in context, Informix's Online Dynamic Server, vintage 1996, used exactly the same multithreaded model adopted by golang (we had to code the context switches by hand though), and exactly the same execution approach as Couchbase's N1QL.
We started with 32K stack sizes, and the world was happy.

I have tried goroutine pooling, but it is quite the kerfufflle, and doesn't work as well as manually forcing an initial stack growth.
We need a better permanent solution, be it a directive (how can a single directive be a sin, when you are forcing me to use two directives and two bogus functions instead?), prediction based on PC, compiler analysis or a runtime function.

**Comment by roudkerk** [comment]
I agree that average stack size isn't likely to work well for many
situations.

For the RPC servers I have looked at, for each request the RPC framework
spawns one or two trivial background goroutines which will drag down the
average size.  So the main goroutine for each request still needs to resize
its stack.

I wonder whether PGO could be taught to smartly select initial stack sizes:
https://go.dev/doc/pgo.  But adapting at runtime would certainly be more
convenient.

**Comment by richardartoul** [comment]
Just chiming in that I'm still working around this issue in my performance-sensitive applications manually for the same reasons mentioned above, an average over the entire application is not reflective of the needs of my critical path.

**Comment by cdvr1993** [comment]
One of our applications was having close to 10% coming from copystack. The average starting stack size reported by the runtime was 2KB.
<img width="280" alt="Screenshot 2023-06-01 at 10 33 51 AM" src="https://github.com/golang/go/assets/3793727/303676b1-bea8-4665-812a-903662bcd75b">

We were able to inject a static stack size and disable any prediction to avoid override the value (linking to private variables from the runtime). We decided to use 8KB stacks and this is the result:
<img width="563" alt="Screenshot 2023-06-06 at 11 27 32 AM" src="https://github.com/golang/go/assets/3793727/0af633c2-6286-4521-89ce-a29c214440f3">

In terms of memory (histograms, don't have the exact value):
- Before: 8MB StackInUse.
- After: 16MB StackInUse.

Could it be possible for the runtime to expose a safe method for the user to select the start stack size? Similar to debug.SetGCPercent

**Comment by aclements** [comment]
It would be useful if someone still having problems with the stack size estimator could post a benchmark that demonstrates the problem. (Not that we don't believe you, it's just hard to iterate on solutions without something to measure!)

Stack sizing is definitely something we've considered for PGO, but also it seems like something where, anything we could do with PGO, we could probably do on the fly. This would have a slower start, but could also be more adaptive.

**Comment by marcogrecopriolo** [comment]
@aclements sorry - I don't have an issue with the stack size estimator anymore, but that's just because I moved from Couchbase to Ditto, and I am now battling with Rust instead ;-)

**Comment by prattmic** [comment]
A commonly proposed refinement to the existing average stack size adjustment is to track average stack size based on goroutine start PCs or perhaps one or more frames at the point the goroutine is created.

One place where I worry this won't work well are things like gRPC servers. This servers launch one goroutine per request, and some of these requests may require lots of stack space (`FooServer.Complicated()`) while others are cheap (`FooServer.Cheap()`). Ideally, we have a large starting stack for the former, and small for the latter.

As far as I can tell [1], gRPC starts this per-request goroutine before determining which request it is handling. This means that no matter the amount of context the runtime uses prior to starting the goroutine, it won't be able to deterministically differentiate between the different types of requests.

This is rather unfortunate, though perhaps if we did this projects like gRPC would be motivated to refactor and start unique goroutine start PCs for each request type.

[1] I've love to be wrong here, but I believe https://github.com/grpc/grpc-go/blob/f1a1fcd042af436be0eb3ac08ed81e2e29fcde4f/server.go#L1024 is the point the goroutine is started.

**Comment by petermattis** [comment]
> One place where I worry this won't work well are things like gRPC servers. This servers launch one goroutine per request, and some of these requests may require lots of stack space (FooServer.Complicated()) while others are cheap (FooServer.Cheap()). Ideally, we have a large starting stack for the former, and small for the latter.

The first time a stack is resized the runtime would likely have sufficient stack context to know what to resize it to. When I posted this issue originally what I noticed is that it was the growth from 2KB -> 4KB -> 8KB -> 16 KB -> 32KB that was slow. The `growStack` hack to force the stack size to grow directly from 2KB to 32KB completely eliminated the overhead (i.e. I couldn't measure a difference between that approach and starting the stack size at 32KB).

**Comment by doujiang24** [comment]
I have a simple, rough idea:
How about introduce a new env variable, i.e. GOMINSTACKSIZE, to configure the min stack size?
cc @aclements 

Background:
I found runtime.morestack take 7.4% CPU time according the perf flamegraph, in our production system.
After hard coding min stack to 32k, it down to 0%.

Memory is large enough since there isn't much goroutines, so we'd love to configure a large min stack size.

Here are some logs that added in `gcComputeStartingStackSize`:
```
scannedStackSize: 282928, scannedStacks: 264
avg StackSize: 1999
```

**Comment by zigo101** [comment]
@doujiang24 If you can identify which goroutine needs a larger stack size, you can use this trick to init it with a larger size: https://github.com/golang/go/issues/18138#issuecomment-932688277

**Comment by doujiang24** [comment]
@zigo101 Thanks, I had noticed that solution, that should works too.
But I'd prefer a way that do not need to hack the business code, since there might be multiple places and may change in the feature.

**Comment by zigo101** [comment]
IMHO, it is best to [support per-goroutine min/initial stack sizes](https://github.com/golang/go/issues/18138#issuecomment-932698922).

**Comment by pixelherodev** [comment]
> A commonly proposed refinement to the existing average stack size adjustment is to track average stack size based on goroutine start PCs or perhaps one or more frames at the point the goroutine is created.
> 
> One place where I worry this won't work well are things like gRPC servers. 

Can confirm that stack growth still remains very heavy on profiles of gRPC servers on the latest Go release, with the average stack size merged in a while ago AFAIK.
